File: Java/cp/happynumber.java

//  Java program to check a number is a Happy 
//  number or not 
import java.util.Scanner;

class happynumber { 
   
// Utility method to return sum of square of 
// digit of n 
static int numSquareSum(int n) 
{ 
    int squareSum = 0; 
    while (n!= 0) 
    { 
        squareSum += (n % 10) * (n % 10); 
        n /= 10; 
    } 
    return squareSum; 
} 
   
//  method return true if n is Happy number 
static boolean isHappynumber(int n) 
{ 
    int slow, fast; 
   
    //  initialize slow and fast by n 
    slow = fast = n; 
    do
    { 
        //  move slow number 
        // by one iteration 
        slow = numSquareSum(slow); 
   
        //  move fast number 
        // by two iteration 
        fast = numSquareSum(numSquareSum(fast)); 
   
    } 
    while (slow != fast); 
   
    //  if both number meet at 1, 
    // then return true 
    return (slow == 1); 
} 
   
//  Driver code to test above methods 
public static void main(String[] args) 
{ 
    int n ; 
    System.out.println("Enter number you want to check");
    Scanner sc = new Scanner(System.in);
    String inte=sc.next();
     n = Integer.parseInt(inte);

    if (isHappynumber(n)) 
        System.out.println(n +  
        " is a Happy number"); 
    else
        System.out.println(n +  
        " is not a Happy number"); 
} 
} 

File: Python/other/HappyNumber.py
def numSquareSum(num): 
    squareSum = 0; 
    while(num): 
        squareSum += (num % 10) * (num % 10); 
        num = int(num / 10); 
    return squareSum; 
  
def isHappynumber(num): 

    slow = num 
    fast = num
    while(True): 

        slow = numSquareSum(slow)
        fast = numSquareSum(numSquareSum(fast)); 
        if(slow != fast): 
            continue
        else: 
            break 
    return (slow == 1) 
  
# Driver Code 
num = int(input("Enter a number: "))
if (isHappynumber(num)): 
    print(num , "is a Happy number") 
else: 
    print(num , "is not a Happy number") 
 
"""
Time Complexity: O(logN)
Space Complexity: O(logN)

Sample Input:
Enter a number: 13
Sample Output:
13 is a Happy number
"""
